lib/org/chromium/webidl/Diff.js

```diff
@@ -0,0 +1,247 @@
+          if (Array.isArray(right[key])) this.dedupDefinition_(right, key);
+
+          // Perform diff with id object to track path and source.
+          var id = {
+            name: key,
+            path: '',
+            leftSources: left[key] ? left[key].sources : null,
+            rightSources: right[key] ? right[key].sources : null,
+          };
+
+          this.diff_(left[key], right[key], id);
+        }.bind(this));
+        return this.chunks;
+      },
+    },
+    {
+      documentation: `Given an array of definitions with the same name,
+        determine if all definitions are the same. If the definitions are all
+        the same (i.e. they have the same members), the entry is replaced
+        with a reference definition.`,
+      name: 'dedupDefinition_',
+      code: function(defns, key) {
+        var changes = [];
+        var id = { name: key, path: '' };
+
+        this.diff_(defns[key], defns[key], id, changes);
+        foam.assert(changes.length === 0,
+            `Diff: ${key} was defined multiple times.
+            The definitions were not identical!`);
+
+        // Concatenate the list of sources.
+        var sources = defns[key].reduce(function(srcArr, def) {
+          return srcArr.concat(def.sources);
+        }, []);
+
+        // Use first object as reference object.
+        defns[key] = defns[key][0];
+        defns[key].sources = sources;
+      },
+    },
+    {
+      documentation: `Given id (type: { name: String, path: String }),
+        status (type: DiffStatus), and left and right, a DiffChunk is created.
+        If opt_changes is provided, the DiffChunk is placed in opt_changes.
+        Otherwise, it is placed in the array of chunks that is returned.`,
+      name: 'addChunk_',
+      code: function(id, status, left, right, opt_changes) {
+        var chunk = this.DiffChunk.create({
+          definitionName: id.name,
+          status: status,
+          propPath: id.path,
+          leftValue: left,
+          leftSources: id.leftSources,
+          rightValue: right,
+          rightSources: id.rightSources,
+        });
+
+        if (opt_changes) opt_changes.push(chunk);
+        else this.chunks.push(chunk);
+      },
+    },
+    {
+      documentation: `If left and right are of primative types, then a direct
+        comparison will be performed and a DiffChunk is created. If left and
+        right are both arrays, a comparison will be performed between the
+        items. If left and right are objects, it will attempt to recursively
+        traverse properties common to both objects. Traversal ends when a
+        difference is found or there is no more properties to compare.`,
+      name: 'diff_',
+      returns: '',
+      code: function (left, right, id, opt_changes) {
+        if (left === undefined || left === null ||
+            right === undefined || right === null) {
+          if (left !== right)
```

Comment:
My intention is, either `left` or `right` is `null` or `undefined` then diff for that property ends there. 

The reasoning behind this is because if either `left` or `right` is `null` or `undefined`, then no recursion will be needed (it is similar to the simple type logic below, but special casing is needed since `typeof(null) === 'object'`). We know for sure that we cannot diff any further on this property on at least one side.